Add Subscription prober

devguyio opened the pull request:
## Proposed Changes

- Adds a prober for the dispatcher subscriptions status

### TODO
- [x] Proper manual testing
- [x] Lots of cleanup
- [ ] Adding tests
- [ ] Add hash support

devguyio committed:
Add Subscription prober

Signed-off-by: Ahmed Abdalla <aabdelre@redhat.com>

matzew commented on pkg/channel/consolidated/reconciler/controller/controller.go:
```diff
@@ -18,6 +18,20 @@ package controller
+	"fmt"
+	"net/url"
+
+	v12 "knative.dev/eventing/pkg/apis/duck/v1"
```
`v12` ?

matzew replied:
duckv1 ?

matzew commented on pkg/channel/consolidated/reconciler/controller/controller.go:
```diff
@@ -36,19 +50,64 @@ import (
+
+	// Get the Dispatcher Service Endpoints and propagate the status to the Channel
+	// endpoints has the same name as the service, so not a bug.
+	eps, err := endpoints.Get(ctx).Lister().Endpoints(dispatcherNamespace).Get(dispatcherName)
```
eps ?

matzew commented on pkg/channel/consolidated/reconciler/controller/controller.go:
```diff
@@ -36,19 +50,64 @@ import (
+		return nil, fmt.Errorf("failed to get internal service: %w", err)
+	}
+	var readyIPs []string
+
```
remove empty line

matzew commented on pkg/channel/consolidated/reconciler/controller/controller.go:
```diff
@@ -36,19 +50,64 @@ import (
+	return []status.ProbeTarget{
+		{
+			PodIPs:  sets.NewString(readyIPs...),
+			PodPort: "8080", Port: "8080", URLs: uls,
```
`PodPorts` vs `Port` ?

matzew commented on pkg/channel/consolidated/reconciler/controller/controller.go:
```diff
@@ -127,3 +194,8 @@ func NewController(
 	return impl
 }
+
+func NewProbeTargetLister(logger *zap.SugaredLogger, lister corev1listers.EndpointsLister) status.ProbeTargetLister {
```
usage of `lister` ?

matzew commented on pkg/channel/consolidated/reconciler/controller/kafkachannel.go:
```diff
@@ -93,7 +94,7 @@ func newDispatcherServiceWarn(err error) pkgreconciler.Event {
 
 func newServiceAccountWarn(err error) pkgreconciler.Event {
-	return pkgreconciler.NewEvent(corev1.EventTypeWarning, "DispatcherServiceAccountFailed", "Reconciling dispatcher ServiceAccount failed: %s", err)
+	return pkgreconciler.NewEvent(corev1.EventTypeWarning, "Dispatc erServiceAccountFailed", "Reconciling dispatcher ServiceAccount failed: %s", err)
```
typo

matzew commented on pkg/channel/consolidated/status/status.go:
```diff
@@ -0,0 +1,469 @@
+/*
+Copyright 2019 The Knative Authors.
```
update your IDE template :-)

matzew replied:
I now see why that is :-) 


Once this is merged, let's have a GH issue that we, eventually, remove all this and go w/ the `knative/pkg` bits

matzew commented on pkg/channel/consolidated/status/status.go:
```diff
@@ -0,0 +1,469 @@
+	"sync"
+	"time"
+
+	v12 "knative.dev/eventing/pkg/apis/duck/v1"
```
`duckv1` ?

matzew commented on pkg/channel/consolidated/status/status.go:
```diff
@@ -0,0 +1,469 @@
+}
+
+// ProbeTarget contains the URLs to probes for a set of Pod IPs serving out of the same port.
+type ProbeTarget struct {
+	PodIPs  sets.String
+	PodPort string
+	Port    string
+	URLs    []*url.URL
+}
```
as said above, add some comment what that all is / why

matzew commented on pkg/channel/consolidated/status/status.go:
```diff
@@ -0,0 +1,469 @@
+	}
+}
+
+func computeHash(sub v12.SubscriberSpec) ([sha256.Size]byte, error) {
```
`computeSubscriptionHash` ?

devguyio commented:
@matzew thanks for the review, I added a TODO section to show that this still needs proper cleanup (comments, naming, tests etc) , mainly would like a feedback on the approach.

devguyio committed:
Fix endpoints informer in cons. KafkaChannel controller

devguyio committed:
Fix unittests after adding status prober

Signed-off-by: Ahmed Abdalla <aabdelre@redhat.com>

devguyio committed:
Format and order go imports in cons. channel controller

Signed-off-by: Ahmed Abdalla <aabdelre@redhat.com>

devguyio committed:
Rename import alias and remove unused variable

Signed-off-by: Ahmed Abdalla <aabdelre@redhat.com>

slinkydeveloper reviewed (commented):
This is going in the right direction :)

devguyio committed:
Add dispatcher prober test for tesitng a single pod

Signed-off-by: Ahmed Abdalla <aabdelre@redhat.com>

devguyio committed:
Support probing dispatchers for multiple partitions kafka channels

Signed-off-by: Ahmed Abdalla <aabdelre@redhat.com>

matzew commented:
Looking at TLS failures.  Is the prober able to access a secured instance of apache kafka ?

matzew commented:
@maschmid is it worth to add your test case for this scenario to this PR, or a follow up PR ?

maschmid commented:
> @maschmid is it worth to add your test case for this scenario to this PR, or a follow up PR ?
> 
> eg. link to your testcase for that?

If you mean a testcase for "event can be sent immediately after the channel's subscribtion is Ready, we already have channel_single_event_test in eventing e2e... we just need to get rid of artificial sleeps like the one in

https://github.com/knative/eventing/blob/main/test/lib/operation.go#L108-L110

devguyio committed:
Update deps

Signed-off-by: Ahmed Abdalla <aabdelre@redhat.com>

devguyio committed:
Fix conumer handler test

Signed-off-by: Ahmed Abdalla <aabdelre@redhat.com>

devguyio committed:
remove unused hashes from status probing test

Signed-off-by: Ahmed Abdalla <aabdelre@redhat.com>

lberk commented on pkg/channel/consolidated/reconciler/controller/kafkachannel_test.go:
```diff
@@ -334,8 +332,7 @@ func TestAllCases(t *testing.T) {
-			consumerGroupWatcher: NewConsumerGroupWatcher(ctx, &FakeClusterAdmin{}, 100*time.Millisecond),
-			kafkachannelLister:   listers.GetKafkaChannelLister(),
+			kafkachannelLister: listers.GetKafkaChannelLister(),
 			// TODO fix
```
is this TODO still valid?

devguyio replied:
I'm honestly unaware of the context of this TODO. It's been there from a much earlier change that used the Kafka Channel informer to enqueue on any update

lberk commented on pkg/channel/consolidated/reconciler/controller/kafkachannel.go:
```diff
@@ -618,8 +582,9 @@ func (r *Reconciler) FinalizeKind(ctx context.Context, kc *v1beta1.KafkaChannel)
-	if r.consumerGroupWatcher != nil {
-		r.consumerGroupWatcher.Forget(string(kc.ObjectMeta.UID))
-	}
+	//TODO
```
TODO?

lberk commented on pkg/channel/consolidated/dispatcher/dispatcher.go:
```diff
@@ -291,13 +292,18 @@ func (d *KafkaDispatcher) UpdateHostToChannelMap(config *Config) error {
+		sub,
+		d.dispatcher,
+		d.channelSubscriptions[channelRef],
+		groupID,
```
Do we really need to pass the `groupID`? it looks like it can be constructed by all the parameters passed to the `consumerMessageHandler{}`

devguyio replied:
it's because `consumerMessageHandler` shouldn't presume how the ConsumerGroup gets calculated. here it's like this, in KafkaSource (which reuses the `consumeMessageHandler`) does something different.

lberk commented on pkg/channel/consolidated/reconciler/controller/controller.go:
```diff
@@ -125,5 +147,24 @@ func NewController(
 		Handler:    controller.HandleAll(grCh),
 	})
 
+	podInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
```
Given the lack of e2e currently, this logical group in particular might be good to add unit tests to

devguyio replied:
do you mind elaborating more? I'm not sure exactly what should such unit tests do? when pods are deleted, that `CancelPodProbing` gets called?

lberk replied:
you can inject objects into the fake informers, and ensure that the filterfunc's are behaving as expected

devguyio replied:
I think that's pretty much covered here? https://github.com/knative/pkg/blob/main/reconciler/filter_test.go

lberk replied:
I'm talking about the specific number and type of objects we'll have before, and then after a resync given the number of specific handlers and filterfuncs we have.  This part of the diff has a lot of codecov warnings about lack of unit tests.  Is there a particular reason we're looking to avoid sufficient coverage for this block?

matzew replied:
sounds reasonable to get them covered via tests. if possible.

slinkydeveloper commented on pkg/channel/consolidated/reconciler/controller/controller.go:
```diff
@@ -85,6 +97,16 @@ func NewController(
+		NewProbeTargetLister(logger, endpointsInformer.Lister()),
+		func(c v1beta1.KafkaChannel, s eventingduckv1.SubscriberSpec) {
+			logger.Debugf("Ready callback triggered for channel: %s/%s subscription: %s", c.Namespace, c.Name, string(s.UID))
+			impl.EnqueueKey(types.NamespacedName{Namespace: c.Namespace, Name: c.Name})
```
I might be wrong here, but shouldn't we enqueue the subscription (which is the one where we want to flip the ready bit)?

devguyio replied:
our `ReconcileKind` is for KafkaChannel. So we must enqueue that to which will eventually reconcile the `spec.status.subscribers` that will end up creating a `patch` only for the desired subscription.

slinkydeveloper commented on pkg/channel/consolidated/reconciler/controller/controller.go:
```diff
@@ -125,5 +147,24 @@ func NewController(
 		Handler:    controller.HandleAll(grCh),
 	})
 
+	podInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
```
IMO this should live inside the prober logic itself (when listing the endpoints, it should just do the difference)

slinkydeveloper commented on pkg/common/consumer/consumer_handler.go:
```diff
@@ -29,6 +29,7 @@ type KafkaConsumerHandler interface {
 	// The returned error is enqueued in errors channel.
 	Handle(context context.Context, message *sarama.ConsumerMessage) (bool, error)
 	SetReady(ready bool)
+	GetConsumerGroup() string
```
The way this interface evolved is a bit weird, but it's ok we can refactor later

slinkydeveloper reviewed (commented):
Small comments, all addressable later

lberk commented on pkg/channel/consolidated/reconciler/controller/kafkachannel.go:
```diff
@@ -256,29 +250,12 @@ func (r *Reconciler) ReconcileKind(ctx context.Context, kc *v1beta1.KafkaChannel
 	for _, s := range ch.Spec.Subscribers {
-		cg := fmt.Sprintf("kafka.%s.%s.%s", ch.Namespace, ch.Name, s.UID)
-		if Find(cgs, cg) {
+		if r, _ := r.statusManager.IsReady(ctx, *ch, s); r {
```
we've called IsReady but not used the error return value, is this the behaviour we're looking for?

slinkydeveloper commented on pkg/channel/consolidated/status/status.go:
```diff
@@ -0,0 +1,514 @@
+	subscriptionKey := sub.UID
+	logger := logging.FromContext(ctx)
+
+	bytes, err := computeHash(sub)
```
Why computing hash? Just using `sub.UID` + `sub.Generation` should be fine and be unique

lberk commented on pkg/channel/consolidated/status/status.go:
```diff
@@ -0,0 +1,514 @@
+				item.targetStates.readyCount.Inc()
+				return true, nil
+			} else {
+				//TODO return and error if the channel doesn't exist?
```
As chatted, lets add a log line here (perhaps info level?) so that we know wthat we've checked dispatcher foo, and haven't found it on that particular pod.

devguyio replied:
while I was adding the log line, I noticed I actually have a log line just above it (before the if) that's giving all the info needed. so now I'm a bit reluctant to add yet another log line with almost same details. wdyt?

devguyio committed:
Apply review comments and add a prober test

Signed-off-by: Ahmed Abdalla <aabdelre@redhat.com>

devguyio committed:
Remove old comment

Signed-off-by: Ahmed Abdalla <aabdelre@redhat.com>

devguyio committed:
Fix fake status manager

Signed-off-by: Ahmed Abdalla <aabdelre@redhat.com>

devguyio committed:
Return error if IsReady returns an error

Signed-off-by: Ahmed Abdalla <aabdelre@redhat.com>

devguyio commented:
going to introduce a small change to use the set of partition numbers instead of a counter
/hold

devguyio committed:
Change probing to be partition based and fix some corner cases of channel deletion

devguyio commented:
@matzew @lberk @slinkydeveloper PR is ready for review. 
I changed the probing to take into consideration the `set` of partitions. So each dispatcher now exposes which subscriptions are ready and which partitions is it dispatching for. The controller need to find the set of all partitions ready to patch the `status.subscribers`
/unhold

matzew commented:
>  PR is ready for review.

@devguyio I see a lot of test failures...

devguyio reviewed (commented):
/retest

devguyio commented:
I'll take a look at the failing tests.

slinkydeveloper reviewed (dismissed):
For me we're good to go!

/hold for somebody else review

devguyio commented:
/retest

devguyio commented:
There's a problem that manifests when running the tests in parallel, single runs of the failed tests succeed. I'll keep digging in.

matzew commented:
Not sure what changed, but before the last changes the tests did actually pass, in parallel mode too

slinkydeveloper commented on pkg/channel/consolidated/status/status.go:
```diff
@@ -0,0 +1,484 @@
+	if err != nil || !ok {
+		// In case of error, enqueue for retry
+		m.workQueue.AddRateLimited(obj)
+		item.logger.Errorf("Probing of %s failed, IP: %s:%s, ready: %t, error: %v (depth: %d)",
```
Error is too much, Warn is fine

devguyio committed:
Change cleanup logic to clean ready subscriptions only

Signed-off-by: Ahmed Abdalla <aabdelre@redhat.com>

devguyio committed:
Remove cleanup to avaid consumers race

Signed-off-by: Ahmed Abdalla <aabdelre@redhat.com>

devguyio commented:
@matzew @slinkydeveloper there's a race that shows up when cleaning up consumer groups of deleted channels (I added that cleanup as part of this PR) So, I removed that part, and we can tackle it in a separate issue that I'll create since it's a bug that's there regardless of this PR.

slinkydeveloper commented:
/test pull-knative-sandbox-eventing-kafka-integration-test-channel-consolidated

Let's try again to see if test pass consistently

devguyio commented:
/retest all

slinkydeveloper commented:
/test pull-knative-sandbox-eventing-kafka-integration-test-channel-consolidated

devguyio commented:
/test all

slinkydeveloper commented:
/retest

https://prow.knative.dev/view/gs/knative-prow/pr-logs/pull/knative-sandbox_eventing-kafka/433/pull-knative-sandbox-eventing-kafka-integration-test-channel-consolidated/1372902798915538944 some kafka source test failure

slinkydeveloper reviewed (approved):
/lgtm

matzew commented:
/lgtm
/approve

@devguyio can we get backport for 0.20 and 0.21?

devguyio commented:
Yep. Will work on it.
/unhold
